File: java/cours/src/session3/TD/MotRecursif/MotConcat.java
package session3.TD.MotRecursif;

public class MotConcat implements Mot{
	private Mot gauche;
	private Mot droit;

	public MotConcat(Mot g, Mot d) {
		this.gauche = g;
		this.droit = d;
	}

	@Override
	public int taille() {
		return this.gauche.taille() + this.droit.taille();
	}

	@Override
	public boolean estConcat() {
		return true;
	}
	
	@Override
	public char premiereLettre() {
		if(this.estVide()){
			throw new UnsupportedOperationException();
		}
		return this.gauche.estVide() ? this.droit.premiereLettre() : this.gauche.premiereLettre();
	}
	
	@Override
	public Mot suite() {
		if(this.estVide()){
			throw new UnsupportedOperationException();
		}
		return this.gauche.estVide() ? this.droit.suite() : new MotConcat(this.gauche.suite(), this.droit);
	}

	@Override
	public Mot gauche() {
		return this.gauche;
	}

	@Override
	public Mot droit() {
		return this.droit;
	}

	@Override
	public String getValue() {
		return this.gauche.getValue() + this.droit.getValue();
	}
}

File: typescript/session3/MotRecursif/MotConcat.ts
import { Mot } from "./mot";
import { Char } from "../Char";

export class MotConcat extends Mot {
  private motGauche: Mot;
  private motDroit: Mot;

  public constructor(g: Mot, d: Mot) {
    super();
    this.motGauche = g;
    this.motDroit = d;
  }

  public taille(): number {
    return this.motGauche.taille() + this.motDroit.taille();
  }

  public estConcat(): boolean {
    return true;
  }

  public premiereLettre(): Char {
    if (this.estVide()) {
      throw new Error("Unsupported exception");
    }
    if (this.motGauche.estVide) return this.motDroit.premiereLettre();
    return this.motGauche.premiereLettre();
  }

  public suite(): Mot {
    if (this.estVide()) {
      throw new Error("Unsupported exception");
    }
    if (this.motGauche.estVide) return this.motDroit.suite();
    return new MotConcat(this.motGauche.suite(), this.motDroit)
  }

  public gauche(): Mot {
    return this.motGauche;
  }

  public droit(): Mot {
    return this.motDroit;
  }

  public getValue(): string {
    return this.motGauche.getValue().concat(this.motDroit.getValue());
  }
}